### PR TITLE
8291897: TerminatingThreadLocal(s) not registered from virtual thread(s)

### PR DESCRIPTION
**Pull request by plevart**
This is an attempt to fix inconsistent behavior of TerminatingThreadLocal(s) when used internally in JDK for per-carrier-thread caches of native ByteBuffer(s) and NativeBuffer(s). If used from virtual thread, such TerminatingThreadLocal(s) are not registered with the carrier thread for the cleanup.
The fix introduces an internal CarrierThreadLocal subclass of ThreadLocal which rewires its API methods to use current thread's carrier thread as the source of ThreadLocalMap instead of current thread itself (if it is a VirtualThread for example). TerminatingThreadLocal is now a subclass of CarrierThreadLocal. It seems only per-carrier-thread caching is a usecase for it. The uses of JavaLangAccess in various places to access a carrier-thread value of given ThreadLocal has been replaced by public API calls and the use of CarrierThreadLocal instead of plain ThreadLocal. JavaLangAccess is still used to dispatch the CarrierThreadLocal API methods, but only for that. Would someone be tempted to use JavaLangAccess methods directly, they now require CarrierThreadLocal argument to guard against missuses.
The REGISTRY of TerminatingThreadLocal(s) that tracks which of them have values bound to a particular carrier thread is now implemented conveniently  with a CarrierThreadLocal.
The test is expanded with a case that demonstrates a situation where a carrier thread is terminated. Since it must wait for 30 seconds for that to happen, only one of the test cases is performed in this mode. The correct logic of TerminatingThreadLocal is still verified with all test-cases using platform threads that can be terminated more rapidly.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291897](https://bugs.openjdk.org/browse/JDK-8291897): TerminatingThreadLocal(s) not registered from virtual thread(s)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9743/head:pull/9743` \
`$ git checkout pull/9743`

Update a local copy of the PR: \
`$ git checkout pull/9743` \
`$ git pull https://git.openjdk.org/jdk pull/9743/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9743`

View PR using the GUI difftool: \
`$ git pr show -t 9743`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9743.diff">https://git.openjdk.org/jdk/pull/9743.diff</a>

</details>

### CODE REVIEW

**Review by AlanBateman — COMMENTED** [review]
I did a quick pass over this and it looks good. Adding CarrierThreadLocal is a good idea and as it allows the use-sites to be revert/cleaned up. Just a few minor comments so far.

File: src/java.base/share/classes/sun/nio/ch/Util.java [hunk 1]
```diff
@@ -230,7 +230,7 @@ public static ByteBuffer getTemporaryDirectBuffer(int size) {
             return ByteBuffer.allocateDirect(size);
         }
 
-        BufferCache cache = JLA.getCarrierThreadLocal(bufferCache);
+        BufferCache cache = bufferCache.get();
```
[the remark this answers is not in the document]
Comment (plevart) [reply src/java.base/share/classes/sun/nio/ch/Util.java:233]: Right, I forgot to remove this one.

File: test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java [hunk 2]
```diff
@@ -26,39 +26,77 @@
...
 /*
  * @test
  * @bug 8202788
  * @summary TerminatingThreadLocal unit test
  * @modules java.base/jdk.internal.misc
- * @run main TestTerminatingThreadLocal
+ * @requires vm.continuations
+ * @compile --enable-preview -source ${jdk.version} TestTerminatingThreadLocal.java
+ * @run main/othervm --enable-preview -Djdk.virtualThreadScheduler.parallelism=1 -Djdk.virtualThreadScheduler.maxPoolSize=2 TestTerminatingThreadLocal
  */
```
[the remark this answers is not in the document]
Comment (plevart) [reply test/jdk/jdk/internal/misc/TerminatingThreadLocal/TestTerminatingThreadLocal.java:40]: Done in next commit.

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**